crates/agent/src/evolution.rs

```diff
@@ -0,0 +1,529 @@
+    // until a future commit.
+
+    Ok(changed_collections)
+}
+
+fn validate_evolving_collections(
+    spec_rows: &Vec<DraftSpecRow>,
+    evolving_collections: &BTreeMap<String, String>,
+) -> anyhow::Result<()> {
+    let collection_name_regex = models::Collection::regex();
+    let mut seen = BTreeSet::new();
+    for row in spec_rows
+        .iter()
+        .filter(|r| evolving_collections.contains_key(r.catalog_name.as_str()))
+    {
+        let old_name = row.catalog_name.as_str();
+        seen.insert(old_name.to_owned());
+
+        // Validate that the new collection name is a valid catalog name.
+        // This results in a better error message, since an invalid name could
+        // otherwise result in a database error due to a constraint violation.
+        let new_name = evolving_collections
+            .get(old_name)
+            .expect("already checked contains_key");
+        if !collection_name_regex.is_match(new_name.as_str()) {
+            anyhow::bail!("requested collection name '{new_name}' is invalid");
+        }
+
+        // Validate that the user has the admin capability for any collection
+        // they're requesting to re-create. We do the permission validations
+        // first so that we don't accidentally leak information about
+        // unauthorized entities in the error messages. Note that technically,
+        // you don't need `admin` on the _old_ collection name, as long as
+        // you have it on the new name. But it seems like a good precedent to
+        // require `admin`, since in the future we may wish to somehow mark the
+        // old collection as deprecated.
+        if row.user_capability != Some(Capability::Admin) {
+            anyhow::bail!("you are not authorized to 'admin' the spec '{old_name}'");
+        }
+
+        // We have to validate that the collection has not been deleted in the draft because
+        // deletions cannot currently be represented in a `models::Catalog`. But other validations
+        // of the type will be handled later, when we try to re-name the collections.
+        if row.draft_type.is_none() {
+            anyhow::bail!(
+                "cannot re-create collection '{old_name}' which was already deleted in the draft"
+            );
+        }
+        // This validation isn't technically necessary. Nothing will break if we re-create a collection
+        // that only exists in the draft. But it seems likely to be unintentional, so probably good to
+        // error out here.
+        if row.live_spec_id.is_none() {
+            anyhow::bail!(
+                "cannot re-create collection '{old_name}' because it has never been published"
+            )
+        }
+
+        // Technically, these checks duplicate validations that would already be performed if the
+        // user tries to publish the draft after re-creating the collections. But this allows us
+        // to surface the error more quickly.
+        match row.live_type {
```

Comment:
noting this as a caution area, where we could inadvertently leak information about a spec that the user isn't entitled to see. especially if the "you're a draft admin" check is relaxed above. I would probably seek to filter joined live specs at query time.